feat: api for block metadata including index_dictionary

navinkarkera opened the pull request:
<!--

🫒🫒
🫒🫒🫒🫒         🫒 Note: the Olive master branch has been created.  Please consider whether your change
    🫒🫒🫒🫒     should also be applied to Olive. If so, make another pull request against the
🫒🫒🫒🫒         open-release/olive.master branch, or ping @nedbat for help or questions.
🫒🫒

🌰🌰🌰🌰🌰🌰     🌰 Note: the Nutmeg release is still supported.
                  Please consider whether your change should be applied to Nutmeg as well.

Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply. You may link to information rather than copy it.
More details about the template are at https://github.com/openedx/open-edx-proposals/pull/180
(link will be updated when that document merges)
-->

## Description

Adds new api to return block metadata which includes [index_dictionary](https://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/edx_search.html#which-data-gets-indexed).

**Background:** The text representation of xblock is required in course-discovery/taxonomy-connector to be able to fetch skills/tags related to it using third party API like lightcast.

## Supporting information

`Private-ref`: [BB-6879](https://tasks.opencraft.com/browse/BB-6879)

**Sandbox url:** https://pr31273.sandbox.opencraft.hosting/dashboard

## Testing instructions

- Read through the code
- Use the sandbox to test out API
- Login to the instance with the default `staff` credentials
- Navigate to 
- https://pr31273.sandbox.opencraft.hosting/api/courses/v1/block_metadata/block-v1:edX+DemoX+Demo_Course+type@video+block@0b9e39477cf34507a7a48f74be381fdd?include=index_dictionary

## Deadline

"None" if there's no rush, or provide a specific date or event (and reason) if there is one.

openedx-webhooks commented:
Thanks for the pull request, @navinkarkera!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

bradenmacdonald commented on docs/swagger.yaml:
```diff
@@ -1854,6 +1854,10 @@ paths:
                 Returned only if the "student_view_data" input parameter contains
                 this block's type.
 
+              * index_dictionary: (dict) The index_dictionary JSON data for this block.
```
Should we say _what_ `index_dictionary` actually is? Something like "(usually this is the text content of the block, for search indexing or other purposes)"

navinkarkera replied:
@bradenmacdonald Ah, yes. Updated! Thanks.

bradenmacdonald reviewed (commented):
Looks good to me now, and I tested it on the sandbox (thanks!).

Before I approve I just need to ask @ormsbee: Is there any potential concern with adding a new BlockStructureTransformer which stores larger data fields (in this case, index_dictionary which can contain transcripts / html text) ? I seem to remember something about that.

ormsbee commented:
@bradenmacdonald:

> Before I approve I just need to ask @ormsbee: Is there any potential concern with adding a new BlockStructureTransformer which stores larger data fields (in this case, index_dictionary which can contain transcripts / html text) ? I seem to remember something about that.

Ah, I think you're right. I think block transformers collect all their data into what's essentially one big course-level file blob that is then manipulated in memory for the various transformations. But that means that if `index_dictionary` is large (as in the case of the combined weight of transcripts), it could potentially blow out the size of the entire structure and add overhead to everything using Block Structures–which is a lot of stuff. That would especially degrade performance if it starts exceeding the 1 MB default memcached limit (meaning it always has to go to S3 or equivalent).

FYI @openedx/perf-interest

bradenmacdonald commented:
Ah, darn. Well then @navinkarkera you may need to create a new separate API to get index_dictionary.
The thing is, the course blocks API is designed to serve XBlock metadata _fast_ from cache. But for index_dictionary and your use case, it doesn't really need to be fast, it just needs to be accessible. So it's fine if the new API just initializes the XBlock, reads the index_dictionary, and returns the data, even though that will be slower than the course blocks API. What do you think? Sorry that I didn't see that coming sooner.

@ormsbee Thanks for the quick input here.

bradenmacdonald commented:
FWIW here's how I built a similar API for the "new" XBlock runtime. You won't be able to use the same code but I think the overall approach works: https://github.com/openedx/edx-platform/blob/0afdda3b9a2184cfd2526f7877837402fbbadc0c/openedx/core/djangoapps/xblock/rest_api/views.py#L56-L82

navinkarkera commented:
> Ah, darn. Well then @navinkarkera you may need to create a new separate API to get index_dictionary. The thing is, the course blocks API is designed to serve XBlock metadata _fast_ from cache. But for index_dictionary and your use case, it doesn't really need to be fast, it just needs to be accessible. So it's fine if the new API just initializes the XBlock, reads the index_dictionary, and returns the data, even though that will be slower than the course blocks API. What do you think? Sorry that I didn't see that coming sooner.
> 

@bradenmacdonald No worries. Thanks for the explanation.

navinkarkera committed:
feat: block metadata api - index_dictionary

Adds new api to return block metadata which includes index_dictionary.
Reason for new api instead of adding it to course blocks API: data like
index_dictionary are too large for the cache used by course/blocks
transformers API.

navinkarkera commented:
@bradenmacdonald Added a new API in course blocks to return metadata. For now it returns `index_dictionary`. Please take a look. If it looks good, I can update swagger docs as well.

bradenmacdonald commented on lms/djangoapps/course_api/blocks/views.py:
```diff
@@ -337,6 +339,65 @@ def list(self, request, hide_access_denials=False):  # pylint: disable=arguments
+    **Use Case**
+
+        Returns the block metadata. Large data like index_dictionary related to
+        a block should be fetched using this API.
```
Maybe explicitly add: ", because they are too large for the cache used by the course blocks/transformers API."

bradenmacdonald commented on lms/djangoapps/course_api/blocks/views.py:
```diff
@@ -337,6 +339,65 @@ def list(self, request, hide_access_denials=False):  # pylint: disable=arguments
+        except InvalidKeyError:
+            raise ValidationError(f"'{str(usage_key_string)}' is not a valid usage key.")  # lint-amnesty, pylint: disable=raise-missing-from
+
+        if not has_access(request.user, "staff", usage_key):
```
I think this could also be just user access, not staff access. But it's fine to leave as staff-only for now.

navinkarkera replied:
Had to select `staff` role as that is the only option available for `UsageKey`.
https://github.com/openedx/edx-platform/blob/dccb792f739032183e15c870fabd061ec45c9f78/lms/djangoapps/courseware/access.py#L603-L618

bradenmacdonald commented on lms/djangoapps/course_api/blocks/views.py:
```diff
@@ -337,6 +339,65 @@ def list(self, request, hide_access_denials=False):  # pylint: disable=arguments
+
+        A dictionary of block_id, block_type and keys in "include" param.
+
+        * index_dictionary: (dict) The index_dictionary JSON data
```
Nit: this comment would be more readable to me if it was formatted like this:

```
    **Response Values**
        A dictionary containing:
        * block_id (string)
        * block_type (string)
        * index_dictionary: (dict) The index_dictionary JSON data...
```

bradenmacdonald commented on lms/djangoapps/course_api/blocks/tests/test_views.py:
```diff
@@ -491,3 +491,97 @@ def test_completion_all_course_with_nav_depth(self):
+                settings.CACHE_MIDDLEWARE_SECONDS
+            )
+        else:
+            assert response.get('Cache-Control', None) is None
```
The `cacheable` parameter doesn't seem to be used. So you should either use it in the `test_block_metadata_response` or delete the parameter + the six lines of code above.

bradenmacdonald commented on lms/djangoapps/course_api/blocks/tests/test_views.py:
```diff
@@ -491,3 +491,97 @@ def test_completion_all_course_with_nav_depth(self):
+
+class TestBlockMetadataView(SharedModuleStoreTestCase):  # pylint: disable=test-inherits-tests
+    """
+    Test class for BlocksInCourseView
```
for `BlockMetadataView`

bradenmacdonald reviewed (approved):
Thanks for those changes, @navinkarkera. This looks good to go!

Please address the nits, rebase and get the build green, and I can merge.

:+1:

- [x] I tested this: on the sandbox
- [x] I read through the code
- [x] I checked for accessibility issues: n/a
- [x] Includes documentation: docstrings yeah

navinkarkera commented on docs/swagger.yaml:
```diff
@@ -1660,6 +1660,56 @@ paths:
         in: path
         required: true
         type: string
+  /courses/v1/block_metadata/{usage_key_string}:
+    get:
+      operationId: courses_v1_block_metadata_read
+      summary: '**Use Case**'
+      description: |-
+        Returns the block metadata. Data like index_dictionary related to a
+            block should be fetched using this API, because they are too large for
+            the cache used by the course blocks/transformers API.
```
I had to manually add this part (copy and paste after generating) as running `make swagger` introduces a lot of changes to this file unrelated to this MR.

bradenmacdonald commented on docs/swagger.yaml:
```diff
@@ -1660,6 +1660,56 @@ paths:
+          in: query
+          description: Number of results to return per page.
+          required: false
+          type: integer
```
I was going to merge this just now, but I noticed this minor issue with the docs... there are no `page` and `page_size` parameters right? The only query parameter is `include`

navinkarkera replied:
@bradenmacdonald  Ohh yes, swagger is adding page and page_size to list APIs, missed to remove it. Thanks!

navinkarkera commented:
@bradenmacdonald The tests are failing due to some package called `pgpy`. It seems to be the case for [all latest commits](https://github.com/openedx/edx-platform/commits/master) (no relation to this MR).

bradenmacdonald commented:
No problem @navinkarkera. I will just wait until those tests get fixed on the master branch and then we can rebase and merge this. Is that OK or is it urgent to merge?

bradenmacdonald commented:
@navinkarkera I think the pgpy issue should be fixed now. There was a discussion of it on Slack #ask-tcril if you're interested.

navinkarkera commented:
@bradenmacdonald Thanks! The tests are now passing.

openedx-webhooks commented:
@navinkarkera 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/31273&entry.1484655318=https://github.com/navinkarkera&entry.752974735=2022-11-09+16:11&entry.1917517419=2022-11-29+18:49&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
